File: herminedb-driver/src/main/java/org/hermine/internal/db/driver/OperationGroupImpl.java
/*
 * Copyright 2018 HermineDB's author : Frédéric Montariol,
 * and explicitly declared author of this file if provided.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.hermine.internal.db.driver;

import jdk.incubator.sql2.*;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Scheduler;

import java.time.Duration;
import java.util.concurrent.CompletionStage;
import java.util.function.Consumer;
import java.util.logging.Logger;
import java.util.stream.Collector;

class OperationGroupImpl<S, T> extends AbstractOperation<T> implements OperationGroup<S, T> {

    static final Collector DEFAULT_COLLECTOR = Collector.of(
            () -> null,
            (a, v) -> {},
            (a, b) -> null,
            a -> null);

    static final Mono<Boolean> DEFAULT_CONDITION = Mono.just(true);

    private Mono<Boolean> condition = DEFAULT_CONDITION;

    private Object accumulator;
    private Collector collector;

    /**
     * completed when this OperationGroup is no longer held. Completion of this
     * OperationGroup depends on held.
     *
     * @see submit, releaseProhibitingMoreOperations, submitHoldingForMoreOperations
     */
    private final Mono<S> held;

    /**
     * predecessor of all member Operations and the OperationGroup itself
     */
    private final Mono head;

    /**
     * The last Mono of any submitted member Operation. Mutable until
     * not isHeld().
     */
    private Mono<S> memberTail;

    protected OperationGroupImpl(ConnectionImpl conn, OperationGroupImpl<? super T, ?> group) {
        super(conn, group);
        held = Mono.empty();
        head = Mono.empty();
        memberTail = head;
        collector = DEFAULT_COLLECTOR;
    }

    @Override
    public OperationGroupImpl<S, T> parallel() {
        return this;
    }

    @Override
    public OperationGroupImpl<S, T> independent() {
        return this;
    }

    @Override
    public OperationGroupImpl<S, T> conditional(CompletionStage<Boolean> condition) {
        return this;
    }

    @Override
    public Submission<T> submitHoldingForMoreMembers() {
        return null;
    }

    @Override
    public Submission<T> releaseProhibitingMoreMembers() {
        return null;
    }

    @Override
    public OperationGroupImpl<S, T> collect(Collector<S, ?, T> c) {
        return this;
    }

    @Override
    public Operation<S> catchOperation() {
        return null;
    }

    @Override
    public OperationGroup<S, T> catchErrors() {
        return null;
    }

    @Override
    public <R extends S> ArrayRowCountOperation<R> arrayRowCountOperation(String sql) {
        return null;
    }

    @Override
    public <R extends S> ParameterizedRowCountOperation<R> rowCountOperation(String sql) {
        return null;
    }

    @Override
    public Operation<S> operation(String sql) {
        return null;
    }

    @Override
    public <R extends S> OutOperation<R> outOperation(String sql) {
        return null;
    }

    @Override
    public <R extends S> ParameterizedRowOperation<R> rowOperation(String sql) {
        return null;
    }

    @Override
    public <R extends S> ParameterizedRowPublisherOperation<R> rowPublisherOperation(String sql) {
        return null;
    }

    @Override
    public <R extends S> MultiOperation<R> multiOperation(String s) {
        return null;
    }

    @Override
    public Operation<TransactionOutcome> endTransactionOperation(Transaction trans) {
        return null;
    }

    @Override
    public CompletionStage<TransactionOutcome> commitMaybeRollback(Transaction trans) {
        return null;
    }

    @Override
    public <R extends S> LocalOperation<R> localOperation() {
        return null;
    }

    @Override
    public OperationGroup<S, T> logger(Logger logger) {
        throw new UnsupportedOperationException("Will use other Logger");
    }

    // Override all Operation methods

    @Override
    public OperationGroupImpl<S, T> onError(Consumer<Throwable> handler) {
        return this;
    }

    @Override
    public OperationGroupImpl<S, T> timeout(Duration minTime) {
        return this;
    }

    @Override
    public SubmissionImpl<T> submit() {
        if ( isImmutable() ) throw new IllegalStateException("TODO");
        accumulator = collector.supplier().get();
        SubmissionImpl<T> sub = super.submit();
        held.subscribe();
        return sub;
    }

    // Internal methods

    SubmissionImpl<S> submit(AbstractOperation<S> op) {
        memberTail = op.attachErrorHandler(op.follows(memberTail, getScheduler()));
        return new SubmissionImpl(this::cancel, memberTail);
    }

    @Override
    Mono<T> follows(Mono<?> predecessor, Scheduler scheduler) {
        return condition.flatMap(cond -> {
                    if (cond) {
                        head.thenReturn(predecessor);
                        return held.then(memberTail.map(t -> (T)collector.finisher()
                                        .apply(accumulator))
                                .publishOn(scheduler)
                        );
                    }
                    else {
                        return Mono.empty();
                    }
                }
        );
    }
}

File: herminedb-driver/src/main/kotlin/org/hermine/internal/db/driver/OperationGroupImpl.kt
/*
 * Copyright 2018 HermineDB's author : Frédéric Montariol. Use of this source code is governed by the Apache 2.0 license.
 */

package org.hermine.internal.db.driver

import jdk.incubator.sql2.*
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Job
import kotlinx.coroutines.future.asDeferred
import mu.KotlinLogging
import java.time.Duration
import java.util.concurrent.CompletionStage
import java.util.function.*
import java.util.function.Function
import java.util.logging.Logger
import java.util.stream.Collector

// Place definition above class declaration to make field static
private val logger = KotlinLogging.logger {}

private val DEFAULT_COLLECTOR: Collector<*, *, *> = Collector.of(
        Supplier { null },
        BiConsumer { _, _: Any? -> },
        BinaryOperator { _, _ -> null },
        Function { _: Nothing? -> null })

private val DEFAULT_CONDITION = CompletableDeferred(true)

internal open class OperationGroupImpl<S, T> : AbstractOperation<T>, OperationGroup<S, T> {

    // used only by Session. Will break if used by any other class.
    constructor() : super()

    constructor(session: SessionImpl, group: OperationGroupImpl<T, *>) : super(session, group)

    protected var condition: Deferred<Boolean> = DEFAULT_CONDITION

    lateinit var submission: SubmissionImpl<T>
    var accumulator: Any? = null

    var isIndependent: Boolean = false
    var isParallel: Boolean = false
    protected var collector: Collector<*, *, *> = DEFAULT_COLLECTOR

    /**
     * Completed when this OperationGroup is no longer held. Completion of this
     * OperationGroup depends on held.
     *
     * @see #submit,
     * @see #releaseProhibitingMoreMembers,
     * @see #submitHoldingForMoreMembers
     */
    private val held = CompletableDeferred<Unit>()

    private val parentJob = Job()

    /**
     * Last operation. Mutable until not isHeld().
     */
    protected var lastOp: Deferred<S?>? = null

    override fun parallel(): OperationGroupImpl<S, T> {
        if (isImmutable() || isParallel) throw IllegalStateException("TODO")
        isParallel = true
        return this
    }

    override fun independent(): OperationGroupImpl<S, T> {
        if (isImmutable() || isIndependent) throw IllegalStateException("TODO")
        isIndependent = true
        return this
    }

    override fun conditional(condition: CompletionStage<Boolean>?): OperationGroupImpl<S, T> {
        if (isImmutable() || condition == null) throw IllegalStateException("TODO")
        this.condition = condition.asDeferred()
        return this
    }

    override fun collect(c: Collector<S, *, T>?): OperationGroupImpl<S, T> {
        if (isImmutable() || collector !== DEFAULT_COLLECTOR) throw IllegalStateException("TODO")
        if (c == null) throw IllegalArgumentException("TODO")
        collector = c
        return this
    }

    override fun catchOperation(): AbstractOperation<S> {
        if (!isHeld()) throw IllegalStateException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun <R : S> arrayRowCountOperation(sql: String?): ArrayRowCountOperation<R> {
        if (!isHeld()) throw IllegalStateException("TODO")
        if (sql == null) throw IllegalArgumentException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun <R : S> rowCountOperation(sql: String?): ParameterizedRowCountOperation<R> {
        if (!isHeld()) throw IllegalStateException("TODO")
        if (sql == null) throw IllegalArgumentException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
//        return ParameterizedRowCountOperationImpl(session, this, sql)
    }

    override fun operation(sql: String?): AbstractOperation<S> {
        if (!isHeld()) throw IllegalStateException("TODO")
        if (sql == null) throw IllegalArgumentException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun <R : S> outOperation(sql: String?): OutOperation<R> {
        if (!isHeld()) throw IllegalStateException("TODO")
        if (sql == null) throw IllegalArgumentException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun <R : S> rowOperation(sql: String?): ParameterizedRowOperation<R> {
        if (!isHeld()) throw IllegalStateException("TODO")
        if (sql == null) throw IllegalArgumentException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun <R : S> rowPublisherOperation(sql: String?): ParameterizedRowPublisherOperation<R> {
        if (!isHeld()) throw IllegalStateException("TODO")
        if (sql == null) throw IllegalArgumentException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun <R : S> multiOperation(sql: String?): MultiOperation<R> {
        if (!isHeld()) throw IllegalStateException("TODO")
        if (sql == null) throw IllegalArgumentException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun endTransactionOperation(trans: TransactionCompletion): AbstractOperation<TransactionOutcome> {
        if (!isHeld()) throw IllegalStateException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun <R : S> localOperation(): LocalOperation<R> {
        if (!isHeld()) throw IllegalStateException("TODO")
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun logger(logger: Logger): OperationGroup<S, T> {
        throw UnsupportedOperationException("Will use other Logger")
    }

    // Override Operation methods

    override fun timeout(minTime: Duration): OperationGroupImpl<S, T> {
        super.timeout(minTime)
        return this
    }

    override fun onError(handler: Consumer<Throwable>): OperationGroupImpl<S, T> {
        super.onError(handler)
        return this
    }

    override fun submit(): SubmissionImpl<T> {
        logger.debug { "OperationGroupImpl#submit" }
        if (isImmutable()) throw IllegalStateException("TODO")
        accumulator = collector.supplier().get()
        held.complete(Unit)
        // start all operations
        startOpJobs()
        return super.submit()
    }

    override fun close() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    // Internal methods

    internal fun submitOperation(op: AbstractOperation<S>): SubmissionImpl<S> {
        logger.debug { "OperationGroupImpl#submitOperation" }
        val newLastOp = op.follows(lastOp, context, isParallel, isIndependent, parentJob)
        lastOp = newLastOp
        return SubmissionImpl(this::cancel, newLastOp)
    }

    override suspend fun operate(): T? {
        logger.debug { "OperationGroupImpl#operate" }
        return if (condition.await()) {
            held.join() // wait until submit or releaseProhibitingMoreMembers

            (collector.finisher() as Function<Any?, Any>)
                    .apply(accumulator) as T
        } else {
            null
        }
    }

    fun isHeld() = !held.isCompleted

    private fun startOpJobs() {
        for ((index, opJob) in parentJob.children.withIndex()) {
            val start = opJob.start() // true if job was not started, false if job was already started or completed
            logger.debug { "start $index child operation job = $start" }
        }
    }
}
